tools/pubsys/src/kit/publish_kit/mod.rs

```diff
@@ -0,0 +1,187 @@
+    version: Version,
+}
+
+pub(crate) async fn run(args: &Args, publish_kit_args: &PublishKitArgs) -> Result<()> {
+    // If a lock file exists, use that, otherwise use Infra.toml
+    let infra_config = InfraConfig::from_path_or_lock(&args.infra_config_path, false)
+        .context(error::ConfigSnafu)?;
+    trace!("Parsed infra config: {:?}", infra_config);
+    ensure!(infra_config.kit_vendors.is_some(), error::NoVendorsSnafu);
+
+    // Fetch the vendor container registry uri
+    let vendor_registry_uri = infra_config
+        .kit_vendors
+        .as_ref()
+        .unwrap()
+        .get(&publish_kit_args.vendor)
+        .context(error::VendorNotFoundSnafu {
+            name: publish_kit_args.vendor.clone(),
+        })?;
+    info!(
+        "Found vendor container registry at uri: {}",
+        vendor_registry_uri
+    );
+
+    // Auto resolve the expected paths for the metadata file and kit contents archive
+    let kit_path = publish_kit_args.kit_path.as_path();
+    let metadata_file = kit_path.join("kit-metadata.json");
+    let metadata_file = metadata_file.as_path();
+    ensure!(
+        metadata_file.exists(),
+        error::NoMetadataSnafu {
+            path: metadata_file
+        }
+    );
+    let mut platforms = Vec::new();
+    for arch in ["amd64", "arm64"] {
+        let path = kit_path.join(format!("kit-{}.tar", arch));
+        if path.exists() {
+            platforms.push(format!("linux/{}", arch));
+        }
+    }
+    ensure!(
+        !platforms.is_empty(),
+        error::NoArchiveSnafu { path: kit_path }
+    );
+
+    // Fetch a view of the metadata
+    let metadata_blob = read(metadata_file).await.context(error::IoSnafu)?;
+    let metadata: MetadataView =
+        serde_json::from_slice(metadata_blob.as_slice()).context(error::DeserializationSnafu)?;
+    trace!("Parse kit metadata to view: {:?}", metadata);
+
+    info!(
+        "Found local kit with name {} and version {}",
+        metadata.name, metadata.version
+    );
+
+    // We only really need the version and name from the metadata here to build our publish uri
+    let target_uri = format!(
+        "{}/{}:{}",
+        vendor_registry_uri, metadata.name, metadata.version
+    );
+
+    // Use docker to build and push the kit base image
+    info!(
+        "Creating and pushing kit container via docker to {}",
+        &target_uri
+    );
+    let metadata_encoded =
+        base64::engine::general_purpose::STANDARD.encode(metadata_blob.as_slice());
+
+    let docker_file = format!(
+        r#"FROM scratch
+ARG TARGETARCH
+
+LABEL {}={}
+
+ADD kit-$TARGETARCH.tar /
+"#,
+        KIT_METADATA_KEY, metadata_encoded
+    );
+    let docker_file_path = kit_path.join("Dockerfile");
+    write(&docker_file_path, docker_file)
+        .await
+        .context(error::IoSnafu)?;
+    let tag = format!("--tag={}", target_uri);
+    let platform = format!("--platform={}", platforms.join(","));
+
+    let result = Command::new("docker")
```

Comment:
I'd prefer to generate as close as possible to the final output in `buildsys` when building the kit, ideally in a form that can be pushed directly.

For example:
* buildsys emits a layer blob and OCI metadata for x86_64
* buildsys emits a layer blob and OCI metadata for aarch64
* pubsys pushes the layer blobs to the target registry so they're available for reference in a manifest
* pubsys generates the manifest and pushes that to the registry to complete the operation

It gets more tricky if we want to support building aarch64 and x86_64 on separate hosts, then bring them together in the manifest. You should consult with @jpculp so this plays nicely with existing automation around container builds for multiple architectures.

E.g. it may be sufficient if all pubsys does push the kit for the target arch and the target version and commit:
```
public.ecr.aws/c2b5m1s5/bottlerocket-core-kit:v1.2.0-abc123-arm64
public.ecr.aws/c2b5m1s5/bottlerocket-core-kit:v1.2.0-abc123-amd64
```

And then we come along in a later phase and generate the combined manifest and push that to the final location:
```
public.ecr.aws/bottlerocket/bottlerocket-core-kit:v1.2.0
```

We can have pubsys do that combining logic too but I think it's best as a separate concern, because we haven't had to do the "bring two architectures together" thing yet in pubsys and it warrants some additional thought.

Reply:
So I am effectively already assuming this but building the multi platform manifest when needed on the pubsys side. There is no need to generate separate metadata here since we are using the docker trick. Its important to note I am not aware of any way to do what you are suggesting in the example using only the docker cli. 

RIght now the assumption flows in this approach are:

* buildsys builds the x86_64 kit and archives it to a archive called kit-amd64.tar
* buildsys builds the aarch64 kit and archives it to a archive called kit-arm64.tar
* I initially though you could have the registry solve the multi-platform part by just pushing each individaul platform separately but i guess you need to do it together to build a multi image.
* pubsys then will detect which platform archives exist and build the image to push

Reply:
Discussed offline, now I understand better how this works and sounds like the right move